File: write/src/test/scala/com/github/jarol/azure/search/spark/sql/connector/write/SearchFieldAnalyzerType.scala
package com.github.jarol.azure.search.spark.sql.connector.write

import com.azure.search.documents.indexes.models.{LexicalAnalyzerName, SearchField}

/**
 * Support trait for test that deal with field analyzers
 */

trait SearchFieldAnalyzerType {

  /**
   * Retrieve the analyzer related to this instance's type from a field
   * @param field Search field
   * @return the field's analyzer related to this instance's type
   */

  def getAnalyzerFrom(field: SearchField): LexicalAnalyzerName

  /**
   * Get the prefix related to this instance's analyzer type
   * @return prefix to be used on configuration creation
   */

  def prefix: String

  /**
   * Create a raw configuration object for a collection of analyzers
   * @param analyzer analyzer to set
   * @param onFields map with keys being analyzer aliases and values being the list of fields on which to set the analyzer
   * @return a raw configuration for multiple analyzers
   */

  final def rawConfigForAnalyzers(
                                   analyzer: LexicalAnalyzerName,
                                   onFields: Map[String, Seq[String]]
                                 ): Map[String, String] = {

    // Base configuration
    val aliasesMap = Map(
      s"${WriteConfig.FIELD_OPTIONS_PREFIX}$prefix${WriteConfig.ALIASES_CONFIG}" -> onFields.keys.mkString(","),
    )

    onFields.foldLeft(aliasesMap) {
      case (map, (alias, fields)) =>

        // Enrich the existing configuration with the current analyzer configuration
        map ++ rawConfigForAnalyzer(
          analyzer,
          alias,
          fields
        )
    }
  }

  /**
   * Create a raw configuration object for a single analyzer
   * <br>
   * The output map will contain the following keys
   *  - analyzer type
   *  - list of fields on which the analyzer should be set
   * @param analyzer analyzer
   * @param alias analyzer alias
   * @param onFields fields on which the analyzer should be set
   * @return a raw configuration for an analyzer
   */

  final def rawConfigForAnalyzer(
                                  analyzer: LexicalAnalyzerName,
                                  alias: String,
                                  onFields: Seq[String]
                                ): Map[String, String] = {

    Map(
      s"${WriteConfig.FIELD_OPTIONS_PREFIX}$prefix$alias.${WriteConfig.TYPE_SUFFIX}" -> analyzer.toString,
      s"${WriteConfig.FIELD_OPTIONS_PREFIX}$prefix$alias.${WriteConfig.ON_FIELDS_SUFFIX}" -> onFields.mkString(",")
    )
  }
}

object SearchFieldAnalyzerType {

  /**
   * Type related to both Search and index analyzer
   */

  case object SEARCH_AND_INDEX extends SearchFieldAnalyzerType {
    override def getAnalyzerFrom(field: SearchField): LexicalAnalyzerName = field.getAnalyzerName
    override def prefix: String = WriteConfig.ANALYZERS_PREFIX
  }

  /**
   * Type related to Search analyzer
   */

  case object ONLY_SEARCH extends SearchFieldAnalyzerType {
    override def getAnalyzerFrom(field: SearchField): LexicalAnalyzerName = field.getSearchAnalyzerName
    override def prefix: String = WriteConfig.SEARCH_ANALYZERS_PREFIX
  }

  /**
   * Type related to index analyzer
   */

  case object ONLY_INDEX extends SearchFieldAnalyzerType {
    override def getAnalyzerFrom(field: SearchField): LexicalAnalyzerName = field.getIndexAnalyzerName
    override def prefix: String = WriteConfig.INDEX_ANALYZERS_PREFIX
  }
}

File: write/src/main/java/com/github/jarol/azure/search/spark/sql/connector/write/SearchFieldAnalyzerType.java
package com.github.jarol.azure.search.spark.sql.connector.write;

import com.azure.search.documents.indexes.models.LexicalAnalyzerName;
import com.azure.search.documents.indexes.models.SearchField;
import com.github.jarol.azure.search.spark.sql.connector.core.schema.SearchFieldAction;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * Types of Search field lexical analyzers.
 * <br>
 * According to the Azure Search API, a string field can define 3 types of lexical analyzers
 * <ul>
 *     <li>for both searching and indexing</li>
 *     <li>only for searching</li>
 *     <li>only for indexing</li>
 * </ul>
 */

public enum SearchFieldAnalyzerType {

    /**
     * Instance for both searching and indexing
     */

    SEARCH_AND_INDEX(
            "searchAndIndex",
            SearchField::setAnalyzerName,
            SearchField::getAnalyzerName
    ),

    /**
     * Instance for only searching
     */

    SEARCH(
            "search",
            SearchField::setSearchAnalyzerName,
            SearchField::getSearchAnalyzerName
    ),

    /**
     * Instance for only indexing
     */

    INDEX(
            "index",
            SearchField::setIndexAnalyzerName,
            SearchField::getIndexAnalyzerName
    );

    private final String description;
    private final BiFunction<SearchField, LexicalAnalyzerName, SearchField> setter;
    private final Function<SearchField, LexicalAnalyzerName> getter;

    /**
     * Create an instance
     * @param description description
     * @param setter function for setting the analyzer
     * @param getter function for getting an analyzer
     */

    @Contract(pure = true)
    SearchFieldAnalyzerType(
            @NotNull String description,
            @NotNull BiFunction<SearchField, LexicalAnalyzerName, SearchField> setter,
            @NotNull Function<SearchField, LexicalAnalyzerName> getter
    ) {
        this.description = description;
        this.setter = setter;
        this.getter = getter;
    }

    /**
     * Get the description of this instance
     * @return instance's description
     */

    @Contract(pure = true)
    public String description() {
        return description;
    }

    /**
     * Get a {@link SearchFieldAction} for setting an analyzer
     * @param name analyzer to set
     * @return an action for setting an analyzer on a field
     */

    public @NotNull SearchField setOnField(
            @NotNull SearchField field,
            @NotNull LexicalAnalyzerName name
    ) {

        return setter.apply(field, name);
    }

    /**
     * Get an analyzer instance from a field
     * @param field a Search field
     * @return an analyzer instance, or null if the analyzer was not set
     */

    public @Nullable LexicalAnalyzerName getFromField(
            @NotNull SearchField field
    ) {
        return getter.apply(field);
    }
}
